File: backend/reservation-service/src/main/java/com/quisin/reservation/repository/ReservationRepository.java
package com.quisin.reservation.repository;

import com.quisin.reservation.model.Reservation;
import com.quisin.reservation.model.ReservationStatus;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDateTime;
import java.util.List;

public interface ReservationRepository extends JpaRepository<Reservation, Long> {
    
    List<Reservation> findByUserId(String userId);
    
    List<Reservation> findByRestaurantId(String restaurantId);
    
    List<Reservation> findByRestaurantIdAndStatus(String restaurantId, ReservationStatus status);
    
    @Query("SELECT r FROM Reservation r WHERE r.restaurantId = :restaurantId " +
           "AND r.reservationTime BETWEEN :startTime AND :endTime")
    List<Reservation> findByRestaurantIdAndTimeRange(
        @Param("restaurantId") String restaurantId,
        @Param("startTime") LocalDateTime startTime,
        @Param("endTime") LocalDateTime endTime
    );
    
    @Query("SELECT COUNT(r) FROM Reservation r WHERE r.restaurantId = :restaurantId " +
           "AND r.reservationTime = :time AND r.status NOT IN ('CANCELLED', 'NO_SHOW')")
    int countActiveReservationsForTimeSlot(
        @Param("restaurantId") String restaurantId,
        @Param("time") LocalDateTime time
    );
    
    @Query("SELECT r FROM Reservation r WHERE r.status = 'PENDING' " +
           "AND r.createdAt < :cutoffTime")
    List<Reservation> findExpiredPendingReservations(
        @Param("cutoffTime") LocalDateTime cutoffTime
    );
    
    boolean existsByRestaurantIdAndTableIdAndReservationTimeAndStatusNot(
        String restaurantId,
        String tableId,
        LocalDateTime reservationTime,
        ReservationStatus status
    );
    
    @Query("SELECT DISTINCT r.tableId FROM Reservation r WHERE r.restaurantId = :restaurantId " +
           "AND r.reservationTime = :time AND r.status NOT IN ('CANCELLED', 'NO_SHOW')")
    List<String> findOccupiedTableIds(
        @Param("restaurantId") String restaurantId,
        @Param("time") LocalDateTime time
    );
} 

File: backend/reservation-service/src/main/kotlin/com/quisin/reservation/repository/ReservationRepository.kt
package com.quisin.reservation.repository

import com.quisin.reservation.domain.Reservation
import com.quisin.reservation.domain.ReservationStatus
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.time.LocalDateTime
import java.util.UUID

@Repository
interface ReservationRepository : JpaRepository<Reservation, UUID> {
    // Find reservations by restaurant
    fun findByRestaurantId(restaurantId: UUID): List<Reservation>

    // Find reservations by customer
    fun findByCustomerId(customerId: UUID): List<Reservation>

    // Find reservations by table
    fun findByTableId(tableId: UUID): List<Reservation>

    // Find reservations by status
    fun findByRestaurantIdAndStatus(restaurantId: UUID, status: ReservationStatus): List<Reservation>

    // Find reservations within a date range
    fun findByRestaurantIdAndReservationDateTimeBetween(
        restaurantId: UUID, 
        startDateTime: LocalDateTime, 
        endDateTime: LocalDateTime
    ): List<Reservation>

    // Find active (confirmed or pending) reservations for a specific table
    @Query("""
        SELECT r FROM Reservation r 
        WHERE r.tableId = :tableId 
        AND r.status IN ('CONFIRMED', 'PENDING') 
        AND r.reservationDateTime BETWEEN :startTime AND :endTime
    """)
    fun findActiveReservationsForTable(
        tableId: UUID, 
        startTime: LocalDateTime, 
        endTime: LocalDateTime
    ): List<Reservation>

    // Count reservations by restaurant and status
    fun countByRestaurantIdAndStatus(restaurantId: UUID, status: ReservationStatus): Long
}
